### ISSUE TITLE
Clarify dataset details + request DATASET.md

### ISSUE THREAD

**Issue by serenity-ruochen** [HarlandZZC/deep_nuclei_parcellation#1]
Hi Haolin @HarlandZZC ,

Thanks for your great work and releasing the code. I’m trying to reproduce the experiments in the paper (DeepNuParc: A novel deep clustering framework for fine-scale parcellation of brain nuclei using diffusion MRI tractography), but I’m still not fully sure about the dataset preparation details from the current instructions.

Could you please clarify (or add a short DATASET.md) covering:

1. Which exact HCP release/package you used (e.g., HCP-YA S1200 vs others) and what files are required.
2. The exact mapping from HCP downloaded folders/files to this repo’s expected input format.
3. The 100 subject IDs and the exact 80/20 train/test split used in the paper.

These details would make it much easier to reproduce your reported results.

**Comment by HarlandZZC** [comment]
Thank you for your interest in our work and for opening the issue. We’re very sorry for the delayed response.

We will organize the relevant details and prepare a short DATASET.md to clarify the dataset preparation process. Once it is ready, we will update the repository and provide a detailed explanation there.

Thank you for your patience and for bringing this up.

Best regards,
Haolin

**Comment by serenity-ruochen** [comment]
Hi Haolin,

Thank you very much for your response. I just run all the steps (with the help of GPT) but the results are far different from the paper.

```
python ./Model/HCP.py --infolder datasets/amygdala_f9000_k100_iter2 --csv f9000_k100_iteration2_label18_54_append_binary_smooth.csv --outfolder ./Model/output_amygdala1 --input_dim 100 --lamda 15000 --beta 0.5 (according to the hyperparameters described in the paper)

python ./Model/visualize_clustering_results.py --subject_folder site_folder --outfolder ./Model/output_amygdala1

python ./Model/compute_dice.py --folder ./Model/output_amygdala1 --n_clusters 9                                     
Average Dice scores:                                                                                                                                                                                   
Class 1: 0.003276253981181193                                                                                                                                                                          
Class 2: 0.22105037093919674                                                                                                                                                                           
Class 3: 0.16125516960816663                                                                                                                                                                           
Class 4: 0.02791027768817312                                                                                                                                                                           
Class 5: 0.037613409807576575                                                                                                                                                                          
Class 6: 0.0822263211693304                                                                                                                                                                            
Class 7: 0.002815057152864011                                                                                                                                                                          
Class 8: 0.0020596630465051515                                                                                                                                                                         
Class 9: 0.002542320714065866

python ./Model/cluster_statistics.py --infolder ./Model/output_amygdala1                                            
Class counts:                                                                                                                                                                                          
Class 1.0: 276 voxels (0.95% of total)                                                                                                                                                                 
Class 2.0: 14494 voxels (49.69% of total)                                                                                                                                                              
Class 3.0: 7285 voxels (24.97% of total)                                                                                                                                                               
Class 4.0: 2344 voxels (8.04% of total)          
Class 5.0: 2577 voxels (8.83% of total)          
Class 7.0: 168 voxels (0.58% of total)           
Class 8.0: 127 voxels (0.44% of total)           
Class 6.0: 1793 voxels (6.15% of total)          
Class 9.0: 107 voxels (0.37% of total)
```


The biggest problem is that **I did not know which step causes the error**... I am guessing the hyperparameters you used for UKF
```bash
MIN_GA=0.06
MIN_FA=0.08
SEED_FA_LIMIT=0.1
SEEDS_PER_VOXEL=1
NUM_THREADS=8
NUM_TENSOR=2

echo "Starting Batch Processing..."

for SUBJ_DIR in "$DATA_ROOT"/sub-*; do
    if [ -d "$SUBJ_DIR" ]; then
        SUBJ_FOLDER_NAME=$(basename "$SUBJ_DIR")
        SUBJ_ID=${SUBJ_FOLDER_NAME#sub-}
        
        echo "----------------------------------------"
        echo "Processing Subject: $SUBJ_ID"

        BASE_DWI_DIR="$SUBJ_DIR/ses-1/dwi"
        DWI_FILE="$BASE_DWI_DIR/${SUBJ_FOLDER_NAME}_ses-1_run-1_dwi.nii.gz"
        BVAL_FILE="$BASE_DWI_DIR/${SUBJ_FOLDER_NAME}_ses-1_run-1_dwi.bval"
        BVEC_FILE="$BASE_DWI_DIR/${SUBJ_FOLDER_NAME}_ses-1_run-1_dwi.bvec"
        MASK_FILE="$BASE_DWI_DIR/corrected_masked/${SUBJ_FOLDER_NAME}_ses-1_run-1_dwi_BrainMask.nii.gz"

        OUT_NAME="${SUBJ_ID}-ukftrack.vtk"
        OUT_PATH="$OUTPUT_DIR/$OUT_NAME"

        if [[ -f "$DWI_FILE" && -f "$MASK_FILE" && -f "$BVAL_FILE" && -f "$BVEC_FILE" ]]; then
            
            EXTRA_PARAMS="--seedsPerVoxel,$SEEDS_PER_VOXEL,--numThreads,$NUM_THREADS,--numTensor,$NUM_TENSOR,--stoppingThreshold,$MIN_GA,--stoppingFA,$MIN_FA,--seedingThreshold,$SEED_FA_LIMIT,--recordFA,--recordTrace"
            
            echo "Running UKF via Python..."
            "$UKF_PYTHON_EXE" "$UKF_PY_SCRIPT" \
                -i "$DWI_FILE" \
                -m "$MASK_FILE" \
                --bvecs "$BVEC_FILE" \
                --bvals "$BVAL_FILE" \
                -o "$OUT_PATH" \
                --params "$EXTRA_PARAMS"

            if [ $? -eq 0 ]; then
                echo "[SUCCESS] Generated: $OUT_NAME"
            else
                echo "[FAIL] UKF crashed for subject $SUBJ_ID"
            fi
        else
            echo "[ERROR] Missing files for $SUBJ_ID"
        fi
    fi
done
```



7. Generate the atlas of pass streamlines **(encounter some problems here)**

I have to change [whitematteranalysis/cluster.py#307 thr](https://github.com/SlicerDMRI/whitematteranalysis/blob/master/whitematteranalysis/cluster.py#L307) from 0.25 to 0.26 so that I can run
```bash
    python ./Segmentation/create_atlas.py --infolder streamlines_folder --outfolder atlas_folder --num_fibers 9000 --num_clusters 100 --num_workers 1
```

8. Choose an atlas to split

    Now, you need to select an atlas folder and split the components of each steamline cluster within this atlas that belong to each subject:

```bash
    python ./Segmentation/split_atlas.py --infolder atlas_folder/atlas_f9000_k100/iteration_00002/remove_outliers --outfolder split_atlas_folder/f9000_k100_iter2
```

11. Generate CSV files for each subject
```bash
    python ./HCP_seg/generate_csv_for_site.py --folder site_folder  --f 9000 --k 100 --iteration 2  --labels 18 54
```

12. Combine the CSV files of multiple subjects into one

    For each nucleus, you need to combine the CSV files from all subjects related to it into a single CSV file:

    ```bash
    python ./HCP_seg/append_csv_for_segmentation.py   --infolder site_folder   --outfolder datasets/amygdala_f9000_k100_iter2   --f 9000 --k 100 --iteration 2   --labels 18 54
    ```
13. Merge and tune CSV of different labels
    ```bash
    python ./Segmentation/merge_and_tune_csv_of_different_labels.py   --infolder datasets/amygdala_f9000_k100_iter2   --outfolder datasets/amygdala_f9000_k100_iter2   --f 9000 --k 100 --iteration 2   --labels 18 54   --binarization 1
    ```
14. Gaussian smoothing
    ```bash
    python ./Segmentation/smooth_the_dataset.py   --in_csv  datasets/amygdala_f9000_k100_iter2/f9000_k100_iteration2_label18_54_append_binary.csv   --out_csv datasets/amygdala_f9000_k100_iter2/f9000_k100_iteration2_label18_54_append_binary_smooth.csv   --num_workers 1
    ```
15. Training the model and visualize the parcellation
 ```bash
    CUDA_VISIBLE_DEVICES=1 python ./Model/HCP.py --infolder datasets/amygdala_f9000_k100_iter2 --csv f9000_k100_iteration2_label18_54_append_binary_smooth.csv --outfolder ./Model/output_amygdala1 --input_dim 100 --lamda 15000 --beta 0.5

python ./Model/visualize_clustering_results.py --subject_folder site_folder --outfolder ./Model/output_amygdala1
    
 ```


The above is my command recording. I would be very grateful if you can suggest some directions to reproduce the results reported in the paper. Thanks again!

**Comment by HarlandZZC** [comment]
Thank you for sharing such a detailed account of your reproduction attempts. Honestly, it is quite difficult for us to pinpoint the root cause just from the outputs and commands alone. 

 I will discuss with my supervisor and, if feasible, try to share some of our intermediate process files and/or reference results during a working day, so you have something concrete to compare against. Please give us a bit of time to organize this.

We will follow up here once we have more to share. Thank you again for your patience.

Best regards,
Haolin

**Comment by serenity-ruochen** [comment]
That would be great! Thank you very much :)

**Comment by HarlandZZC** [comment]
Hi,

So sorry for being late, please check the link https://drive.google.com/drive/folders/17PU22eVO42uj6M_HllfjrLPxxW8ZieB8?usp=drive_link

Haolin

**Comment by serenity-ruochen** [comment]
@HarlandZZC  Thanks for your help!

There might be some issues in the shared link :(  All folders are empty.

**Comment by serenity-ruochen** [comment]
Hi Haolin,

I am trying to reproduce the amygdala parcellation experiment. I used the amygdala labels `18 54`, `k=100`, and final parcel number `c=9`.

However, I found that my reproduced Dice scores are much lower than the reported results. My average Dice values are mostly below 0.1, with the best class around 0.22. I traced the issue back to the atlas construction step. The README uses:

```bash
python ./Segmentation/create_atlas.py \
  --infolder streamlines_folder \
  --outfolder atlas_folder \
  --num_fibers 9000 \
  --num_clusters 100
```

But in my data, the selected amygdala pass fibers are far fewer than 9000 per subject. For example, after running `wm_select_pass_fibers_forSeqDilation_site.py` and transforming to MNI space, I get:

```text
min: 720 pass fibers
max: 5852 pass fibers
```

So I had to reduce `--num_fibers` to `700` to make `wm_cluster_atlas.py` run. This produced `f700_k100` features, but the resulting features appear very sparse. Before smoothing, many amygdala voxels have all-zero streamline-cluster vectors. For example:

```text
sub-100307 label 18: 235 / 534 zero vectors
sub-100307 label 54: 720 / 934 zero vectors
sub-100408 label 54: 386 / 462 zero vectors
```

After smoothing, the zero vectors become 0, but the final Dice scores are still very low.

My whole-brain UKF tractography files contain around 350k-400k streamlines per subject, for example:

```text
sub-100307: 360252 fibers
sub-100408: 402123 fibers
sub-101107: 354852 fibers
sub-101309: 384495 fibers
```

The UKF parameters I used are similar to the filename pattern in the README:

```text
minGA = 0.06
minFA = 0.08
seedFALimit = 0.1
seedsPerVoxel = 1
numTensor = 2
```

I have a few questions:

1. In your amygdala experiment, how many selected amygdala pass fibers did each subject typically have before atlas construction?
2. Were all subjects able to provide at least 9000 pass fibers for `--num_fibers 9000`?
3. What UKF tractography parameters were used to generate the tractography files in the paper, especially `seedsPerVoxel`, seeding/stopping thresholds, mask, and any post-filtering?
4. Did you use the same `region_dilation_sequentially.py` settings as in the repository? In the current script, the dilation structure is:
      ```python
      struct = np.zeros((3,3,3))
      struct[1,1,0] = 0.5
      struct[1,0,1] = 0.5
      struct[0,1,1] = 0.5
      ```
      I am wondering whether the paper used this structure, or a standard 3D dilation structure such as `ndimage.generate_binary_structure(3, 1)`.
5. Is `f=9000, k=100` essential for reproducing the reported amygdala result, or did you also evaluate lower `f` values when some subjects had fewer pass fibers?

Any details about the original tractography density and selected pass-fiber counts would be very helpful. Thanks again for sharing the code 👍

**Comment by HarlandZZC** [comment]
Hi,
please check this link: https://drive.google.com/drive/folders/17PU22eVO42uj6M_HllfjrLPxxW8ZieB8?usp=sharing
Does it work?

Haolin

**Comment by serenity-ruochen** [comment]
Thanks for the sharing! I have a quick question about the `sub-100307` file you shared.

**Is the pass-fiber file you gave me for amygdala or thalamus?**

In my run, using the same subject:

```text
whole-brain UKF fibers: 360252
thalamus pass fibers (10,49): 32878
amygdala pass fibers (18,54): 2538
```

But the file you gave me has:

```text
pass fibers: 15958   streamlines_folder_example/sub-100307_ses-1_run-1_pass_fibers-SeqDilation-mni.vtk
```

**So the number does not match my amygdala/thalamus result.**

Could you also share the intermediate DDSurfer folder for this subject?

```text
site_folder/sub-100307/ses-1/dwi/DDSurfer/
```

I want to compare `wmparc`, `SeqDilation`, and the selected pass fibers directly.

Thanks!

**Comment by serenity-ruochen** [comment]
Hi Haolin,

I just used the ukf tractography file you provided (100307-ukftrack_b3000_fsmask_421a7ad_minGA0.06_minFA0.08_seedFALimit0.1.vtk) to rerun steps 3-7 ( amygdala pass fibers; labels 18,54 ).

The resulting passed fibers are `pass fibers: 5408  streamlines_folder_example/sub-100307_ses-1_run-1_pass_fibers-SeqDilation-mni.vtk`, which is still far fewer than the file you provided `pass fibers: 15958   streamlines_folder_example/sub-100307_ses-1_run-1_pass_fibers-SeqDilation-mni.vtk`

**Comment by HarlandZZC** [comment]
please give me sometime to figure about what happens

**Comment by serenity-ruochen** [comment]
Many thanks 👍
